Mathlib/Algebra/Lie/Basic.lean

```diff
@@ -0,0 +1,1094 @@
+/-
+Copyright (c) 2019 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+
+! This file was ported from Lean 3 source module algebra.lie.basic
+! leanprover-community/mathlib commit dc6c365e751e34d100e80fe6e314c3c3e0fd2988
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.Module.Equiv
+import Mathlib.Data.Bracket
+import Mathlib.LinearAlgebra.Basic
+
+/-!
+# Lie algebras
+
+This file defines Lie rings and Lie algebras over a commutative ring together with their
+modules, morphisms and equivalences, as well as various lemmas to make these definitions usable.
+
+## Main definitions
+
+  * `lie_ring`
+  * `lie_algebra`
+  * `lie_ring_module`
+  * `lie_module`
+  * `lie_hom`
+  * `lie_equiv`
+  * `lie_module_hom`
+  * `lie_module_equiv`
+
+## Notation
+
+Working over a fixed commutative ring `R`, we introduce the notations:
+ * `L →ₗ⁅R⁆ L'` for a morphism of Lie algebras,
+ * `L ≃ₗ⁅R⁆ L'` for an equivalence of Lie algebras,
+ * `M →ₗ⁅R,L⁆ N` for a morphism of Lie algebra modules `M`, `N` over a Lie algebra `L`,
+ * `M ≃ₗ⁅R,L⁆ N` for an equivalence of Lie algebra modules `M`, `N` over a Lie algebra `L`.
+
+## Implementation notes
+
+Lie algebras are defined as modules with a compatible Lie ring structure and thus, like modules,
+are partially unbundled.
+
+## References
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 1--3*](bourbaki1975)
+
+## Tags
+
+lie bracket, jacobi identity, lie ring, lie algebra, lie module
+-/
+
+
+universe u v w w₁ w₂
+
+open Function
+
+-- Porting note: mathlib3 had a `@[protect_proj]` here, but adding `protected` to all the fields
```

Comment:
Since protect_proj is currently not implemented as an attribute, we have no choice but to leave it out for now, but I think the decision of whether to include them in the long-run is a refactor which we might want to decide on after the port. The general consensus seems to be to keep things in sync with mathlib3 as much as possible for the time being